protoc_plugin/lib/indenting_writer.dart

```diff
@@ -118,18 +125,27 @@ class IndentingWriter {
     _suffixes[suffixKey] = text;
   }
 
-  @override
-  String toString() {
+  /// Emit the generated source.
+  ///
+  /// This is safe to call multiple times.
+  String emitSource({required bool format}) {
     if (_suffixes.isNotEmpty) {
-      // TODO: We may want to introduce the notion of closing the writer.
       println('');
       for (final key in _suffixes.keys) {
         println(_suffixes[key]!);
       }
       _suffixes.clear();
     }
 
-    return _buffer.toString();
+    var source = _buffer.toString();
+
+    // We don't always want to format the source (for example, we don't want to
```

Comment:
This is an interesting problem. It would be cool if we had some way of "anchoring" the annotated locations in the source.

Something like https://gcc.gnu.org/onlinedocs/cpp/Line-Control.html#Line-Control-1 ...

@lrhn wdyt?

Reply:
I believe these are for references from .proto files to the generated code. Not sure a dart parser works directly here. (but perhaps via symbol names...)